Add support for Azure Managed Identity

DaveCTurner opened the pull request:
With this commit, if no key or SAS token is supplied for an Azure
repository then Elasticsearch will use the `DefaultAzureCredential`
chain defined in the Azure SDK, which will obtain credentials from the
instance metadata service when running on an Azure VM.

DaveCTurner committed:
Add support for Azure Managed Identity

With this commit, if no key or SAS token is supplied for an Azure
repository then Elasticsearch will use the `DefaultAzureCredential`
chain defined in the Azure SDK, which will obtain credentials from the
instance metadata service when running on an Azure VM.

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

DaveCTurner committed:
Update docs/changelog/111344.yaml

DaveCTurner commented:
This PR is the main production code change, and includes tests of some of the new functionality, but does not complete this work. See https://github.com/elastic/elasticsearch/issues/111346 for a rundown of the overall project.

DaveCTurner commented:
Docs build is now green, not sure why it shows failed above still. And this PR doesn't touch any docs anyway.

DaveCTurner committed:
Merge branch 'main' into 2024/07/26/azure-managed-identity-support

henningandersen commented on modules/repository-azure/licenses/jna-NOTICE.txt:
```diff
@@ -0,0 +1 @@
+ 
```
Nit: the other notice files with no contents are empty, I suggest to remove the newline here.

DaveCTurner replied:
We enforce in our build that all instances of the `NOTICE` file for each library are the same. This is a copy of the already-extant `x-pack/plugin/sql/sql-cli/licenses/jna-NOTICE.txt`.

henningandersen commented on modules/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureStorageServiceTests.java:
```diff
@@ -186,25 +186,18 @@ public void testReinitClientWrongSettings() throws IOException {
             assertThat(client11.getSyncClient().getAccountUrl(), equalTo("https://myaccount1.blob.core.windows.net"));
             assertThat(
                 expectThrows(SettingsException.class, () -> plugin.reload(settings2)).getMessage(),
-                is("Neither a secret key nor a shared access token was set for account [myaccount1]")
```
What happens if there is no access, i.e., also no managed identity setup? I wonder if we need a test for that to demonstrate that it fails in a good enough way?

DaveCTurner replied:
The error is actually pretty descriptive, much better than what we get with the AWS SDK, but yeah you're right we should have a test for that too.

DaveCTurner replied:
See 
f705e34

henningandersen reviewed (commented):
I wonder if we need a test of the error handling too?

DaveCTurner commented on modules/repository-azure/build.gradle:
```diff
@@ -53,9 +57,24 @@ dependencies {
+  api "com.nimbusds:oauth2-oidc-sdk:11.9.1"
   api "jakarta.activation:jakarta.activation-api:1.2.1"
   api "jakarta.xml.bind:jakarta.xml.bind-api:2.3.3"
+  api "net.java.dev.jna:jna-platform:${versions.jna}" // Maven says 5.14.0 but this aligns with the Elasticsearch-wide version
+  api "net.java.dev.jna:jna:${versions.jna}" // Maven says 5.14.0 but this aligns with the Elasticsearch-wide version
```
I didn't see anything in [the JNA changelog](https://github.com/java-native-access/jna/blob/master/CHANGES.md) that indicated an upgrade was required. It's hard to tell really, there's risks both ways.

DaveCTurner committed:
Use HTTP for Azure fixture in FIPS tests

We cannot (easily) override the JVM trust store in FIPS tests. This
commit falls back to using HTTP in these cases.

Closes #111279
Closes #111345
Closes #111307
Closes #111377
Closes #111280

DaveCTurner committed:
Merge branch 'main' into 2024/07/26/azure-managed-identity-support

DaveCTurner committed:
Merge branch '2024/07/29/fips-azure-fixture' into 2024/07/26/azure-managed-identity-support

DaveCTurner committed:
Verify sensible exception message on totally missing credentials

DaveCTurner committed:
Merge branch 'main' into 2024/07/26/azure-managed-identity-support

DaveCTurner committed:
Merge branch 'main' into 2024/07/26/azure-managed-identity-support

DaveCTurner committed:
Override AZURE_POD_IDENTITY_AUTHORITY_HOST to force fast failure

henningandersen reviewed (approved):
LGTM.

mark-vieira commented on x-pack/plugin/snapshot-repo-test-kit/qa/azure/build.gradle:
```diff
@@ -45,3 +46,20 @@ tasks.named("javaRestTest") {
+}
+
+tasks.named("managedIdentityJavaRestTest") {
+  systemProperty 'test.azure.fixture', Boolean.toString(useFixture)
```
FWIW, all of this can go up above in the `register` block and you can get rid of this extra configuration block.

DaveCTurner replied:
Hmm that was the first thing I tried, but it didn't work. I'll try again in a follow-up.

mark-vieira replied:
The behavior _should_ be identical.

DaveCTurner replied:
Hm yeah seems ok now, not sure what I did wrong last time: https://github.com/elastic/elasticsearch/pull/111444

piyushkp commented:
@DaveCTurner how do we use azure repository for snapshot using MSI identity ? if Key/Sas token is not provided then how does the repository authenticate to storage blob using Managed identity ? can you provide clear instructions or sample to connect storage blob using MSI ?

DaveCTurner commented:
Thanks very much for your interest in Elasticsearch @piyushkp.

This appears to be a user question, and we'd like to direct these kinds of things to the [Elasticsearch forum](https://discuss.elastic.co/c/elasticsearch). If you can stop by there, we'd appreciate it. This allows us to use GitHub for verified bug reports, feature requests, and pull requests.
